code/components/extra-natives-five/src/TrackNatives.cpp

```diff
@@ -0,0 +1,186 @@
+{
+	return (point1.x - x) * (point1.x - x) + (point1.y - y) * (point1.y - y) + (point1.z - z) * (point1.z - z);
+}
+
+// This has never changed.
+static constexpr int kMaxTrainTracks = 27;
+
+static int32_t FindClosestTrack(rage::Vector3& position, int8_t* outTrack)
+{
+	*outTrack = -1;
+	float closestDistance = std::numeric_limits<float>::infinity();
+	int closestNode = -1;
+
+	for (int i = 0; i < kMaxTrainTracks; i++)
+	{
+		TrackData* track = getTrainTrack(i);
+
+		// Skip if this track is a nullptr or is currently disabled. The game doesn't check for this.
+		if (!track || !track->m_enabled)
+		{
+			continue;
+		}
+
+		for (int n = 0; n < track->m_nodeCount; n++)
+		{
+			TrackNode node = track->m_nodes[n];
+
+			float Distance = calculateDistance(position, node.m_x, node.m_y, node.m_z);
+
+			if (Distance < closestDistance)
+			{
+				closestNode = n;
+				*outTrack = i;
+				closestDistance = Distance;
+			}
+		}
+	}
+
+	return closestNode;
+}
+
+
+static HookFunction hookFunction([]()
+{
+	MH_Initialize();
+	// add missing enabled check for script created trains.
+	MH_CreateHook(hook::get_call(hook::get_pattern("E8 ? ? ? ? 8B D8 E8 ? ? ? ? 44 8A CF")), FindClosestTrack, NULL);
+	MH_EnableHook(MH_ALL_HOOKS);
+
+	// Prevent game code from constantly setting the trains speed while in moving state if it has the "stopsAtStations" flag enabled from setting the train speed to the tracks max speed while moving.
+	hook::nop(hook::get_pattern("F3 0F 10 75 ? 8B 55"), 5);
+});
+
+static TrackData* getAndCheckTrack(fx::ScriptContext& context, std::string_view nn)
+{
+	int trackIndex = context.GetArgument<int>(0);
+	if (trackIndex < 0 || trackIndex > kMaxTrainTracks)
+	{
+		trace("Invalid track index %i passed to %s\n", trackIndex, nn);
+		context.SetResult(0);
+		return NULL;
+	}
+
+	TrackData* track = getTrainTrack(trackIndex);
+
+	if (!track || *(int*)track == NULL)
+	{
+		trace("Track index %i passed to %s does not exist\n", trackIndex, nn);
+		context.SetResult(0);
+		return NULL;
+	}
+
+	return track;
+}
+
+static InitFunction initFunction([]()
+{
+	fx::ScriptEngine::RegisterNativeHandler("SET_TRACK_MAX_SPEED", [](fx::ScriptContext& context)
+	{
+		int maxSpeed = context.CheckArgument<int>(1);
+
+		if (TrackData* track = getAndCheckTrack(context, "SET_TRACK_MAX_SPEED"))
+		{
+			track->m_speed = maxSpeed;
+		}
+	});
+
+	fx::ScriptEngine::RegisterNativeHandler("GET_TRACK_MAX_SPEED", [](fx::ScriptContext& context)
+	{
+		if (TrackData* track = getAndCheckTrack(context, "GET_TRACK_MAX_SPEED"))
+		{
+			context.SetResult<int>(track->m_speed);
+		}
+	});
+
+	fx::ScriptEngine::RegisterNativeHandler("GET_TRACK_BREAKING_DISTANCE", [](fx::ScriptContext& context)
+	{
+		if (TrackData* track = getAndCheckTrack(context, "GET_TRACK_BREAKING_DISTANCE"))
+		{
+			context.SetResult<int>(track->m_breakDistance);
+		}
+	});
+
+	fx::ScriptEngine::RegisterNativeHandler("SET_TRACK_BREAKING_DISTANCE", [](fx::ScriptContext& context)
```

Comment:
Nit: `break` is not technically the correct word here - it should be `brake` instead. Would appreciate if you could make that switch wherever `break` has been used.